src/corvus.c

```diff
@@ -18,14 +18,33 @@
 #include "timer.h"
 #include "alloc.h"
 
+static pthread_spinlock_t signal_lock;
 static struct context *contexts;
 
-void sig_handler(int sig)
+void sigsegv_handler(int sig)
 {
-    int i;
+    if (pthread_spin_trylock(&signal_lock) == EBUSY) {
+        sigset_t mask;
+        sigemptyset(&mask);
+        pselect(0, NULL, NULL, NULL, NULL, &mask);
```

Comment:
True, `pselect` could suspend the signal for a while.

I'd like to say that I'm not suggesting **the return way** nor **the block way**.

By **the block way**, I mean the way of using signal mask you said before(not `pselect`), both of them just don't feel right IMHO, and I don't see a significant difference in between.

With **the return way** the signal handler will be triggered again and again, just like what you said, but isn't that the same for **the block way**(not to mention that in a multithreaded process you have to use a different way to mask all threads)? Yes, the handler is not called, but it still occurs again and again, the handler is just ignoring it.


---

I'm starting to think that it will be better if Corvus handles `SIGSEGV` only in debuging, and let core dump do its work in production.